File: apps/ml/clients/tmdb.py
import os
import requests
from dotenv import load_dotenv

# Load environment variables
if os.path.exists(".env.local"):
    load_dotenv(".env.local")
else:
    load_dotenv()

TMDB_API_KEY = os.getenv("TMDB_API_KEY")
TMDB_READ_ACCESS_TOKEN = os.getenv("TMDB_API_READ_ACCESS_TOKEN")
TMDB_BASE_URL = "https://api.themoviedb.org/3"


def get_headers():
    """Returns headers for TMDb API requests using Bearer token."""
    return {
        "Authorization": f"Bearer {TMDB_READ_ACCESS_TOKEN}",
        "Content-Type": "application/json"
    }


def search_movie(query: str, language: str = "fr-FR"):
    """Search for a movie by title."""
    url = f"{TMDB_BASE_URL}/search/movie"
    params = {
        "query": query,
        "language": language
    }
    response = requests.get(url, headers=get_headers(), params=params)
    response.raise_for_status()
    return response.json()


def get_movie_details(movie_id: int, language: str = "fr-FR"):
    """Get detailed information about a movie."""
    url = f"{TMDB_BASE_URL}/movie/{movie_id}"
    params = {"language": language}
    response = requests.get(url, headers=get_headers(), params=params)
    response.raise_for_status()
    return response.json()


def get_movie_credits(movie_id: int):
    """Get cast and crew for a movie."""
    url = f"{TMDB_BASE_URL}/movie/{movie_id}/credits"
    response = requests.get(url, headers=get_headers())
    response.raise_for_status()
    return response.json()


def get_popular_movies(language: str = "fr-FR", page: int = 1):
    """Get a list of popular movies."""
    url = f"{TMDB_BASE_URL}/movie/popular"
    params = {"language": language, "page": page}
    response = requests.get(url, headers=get_headers(), params=params)
    response.raise_for_status()
    return response.json()


if __name__ == "__main__":
    # Quick test
    print("Testing TMDb API connection...")
    
    if not TMDB_READ_ACCESS_TOKEN:
        print("ERROR: TMDB_API_READ_ACCESS_TOKEN not found in environment")
        exit(1)
    
    try:
        result = search_movie("Inception")
        if result.get("results"):
            movie = result["results"][0]
            print(f"✓ Connection successful!")
            print(f"  Found: {movie['title']} ({movie.get('release_date', 'N/A')[:4]})")
            print(f"  Vote: {movie['vote_average']}/10")
        else:
            print("Connection OK but no results found.")
    except Exception as e:
        print(f"✗ Connection failed: {e}")

File: apps/web/src/lib/tmdb.ts
const TMDB_BASE_URL = "https://api.themoviedb.org/3";

interface TMDbConfig {
    apiKey: string | undefined;
    accessToken: string | undefined;
}

const config: TMDbConfig = {
    apiKey: process.env.TMDB_API_KEY,
    accessToken: process.env.TMDB_API_READ_ACCESS_TOKEN,
};

export interface TMDbMovie {
    id: number;
    title: string;
    original_title: string;
    overview: string;
    poster_path: string | null;
    backdrop_path: string | null;
    release_date: string;
    vote_average: number;
    vote_count: number;
    popularity: number;
    runtime: number | null;
    budget: number;
    revenue: number;
    tagline: string | null;
    homepage: string | null;
    genre_ids?: number[];
    genres?: { id: number; name: string }[];
}

export interface TMDbCredits {
    id: number;
    cast: {
        id: number;
        name: string;
        character: string;
        profile_path: string | null;
    }[];
    crew: {
        id: number;
        name: string;
        job: string;
    }[];
}

export interface TMDbSearchResponse {
    page: number;
    results: TMDbMovie[];
    total_pages: number;
    total_results: number;
}

function getHeaders(): HeadersInit {
    if (config.accessToken) {
        return {
            Authorization: `Bearer ${config.accessToken}`,
            "Content-Type": "application/json",
        };
    }
    return {
        "Content-Type": "application/json",
    };
}

function buildUrl(endpoint: string, params?: Record<string, string>): string {
    const url = new URL(`${TMDB_BASE_URL}${endpoint}`);

    // Add API key if using key-based auth
    if (!config.accessToken && config.apiKey) {
        url.searchParams.append("api_key", config.apiKey);
    }

    if (params) {
        Object.entries(params).forEach(([key, value]) => {
            url.searchParams.append(key, value);
        });
    }

    return url.toString();
}

export async function searchMovies(
    query: string,
    page = 1,
    language = "fr-FR"
): Promise<TMDbSearchResponse> {
    const url = buildUrl("/search/movie", {
        query,
        page: String(page),
        language,
        include_adult: "false",
    });

    const response = await fetch(url, {
        headers: getHeaders(),
        next: { revalidate: 3600 }, // Cache for 1 hour
    });

    if (!response.ok) {
        throw new Error(`TMDb API error: ${response.status}`);
    }

    return response.json() as Promise<TMDbSearchResponse>;
}

export async function getPopularMovies(page = 1, language = "fr-FR"): Promise<TMDbSearchResponse> {
    const url = buildUrl("/movie/popular", {
        page: String(page),
        language,
    });

    const response = await fetch(url, {
        headers: getHeaders(),
        next: { revalidate: 3600 },
    });

    if (!response.ok) {
        throw new Error(`TMDb API error: ${response.status}`);
    }

    return response.json() as Promise<TMDbSearchResponse>;
}

export async function getMovieDetails(movieId: number, language = "fr-FR"): Promise<TMDbMovie> {
    const url = buildUrl(`/movie/${movieId}`, {
        language
    });

    const response = await fetch(url, {
        headers: getHeaders(),
        next: { revalidate: 86400 }, // Cache for 24 hours
    });

    if (!response.ok) {
        throw new Error(`TMDb API error: ${response.status}`);
    }

    return response.json() as Promise<TMDbMovie>;
}

export async function getMovieCredits(movieId: number, language = "fr-FR"): Promise<TMDbCredits> {
    const url = buildUrl(`/movie/${movieId}/credits`, {
        language
    });

    const response = await fetch(url, {
        headers: getHeaders(),
        next: { revalidate: 86400 },
    });

    if (!response.ok) {
        throw new Error(`TMDb API error: ${response.status}`);
    }

    return response.json() as Promise<TMDbCredits>;
}

export function getImageUrl(
    path: string | null,
    size: "w92" | "w154" | "w185" | "w342" | "w500" | "w780" | "original" = "w500"
): string | null {
    if (!path) return null;
    return `https://image.tmdb.org/t/p/${size}${path}`;
}
